RevenueCatUI/Templates/V2/Components/Carousel/CarouselComponentView.swift

```diff
@@ -0,0 +1,693 @@
+            // Non-loop clamp
+            index = max(0, min(index, data.count - 1))
+        }
+
+        // Pause auto-play for 10 seconds
+        pauseAutoPlay(for: 10)
+    }
+
+    private func pauseAutoPlay(for seconds: TimeInterval) {
+        isPaused = true
+        pauseEndDate = Date().addingTimeInterval(seconds)
+
+        // Restart auto-play after `seconds` seconds
+        DispatchQueue.main.asyncAfter(deadline: .now() + seconds) {
+            if Date() >= self.pauseEndDate! {
+                self.isPaused = false
+            }
+        }
+    }
+
+    // MARK: - Expanding
+
+    private func expandDataIfNeeded() {
+        // If index is in the first copy, prepend another
+        if index < originalCount {
+            let newCopyIndex = lowestCopyIndex() - 1
+            let newItems = makeItems(forCopyIndex: newCopyIndex)
+            data.insert(contentsOf: newItems, at: 0)
+            index += originalCount // keep user in same “visual” position
+        }
+
+        // If index is in the last copy, append another
+        if index >= data.count - originalCount {
+            let newCopyIndex = highestCopyIndex() + 1
+            let newItems = makeItems(forCopyIndex: newCopyIndex)
+            data.append(contentsOf: newItems)
+        }
+    }
+
+    // MARK: - Pruning
+
+    private func pruneDataIfNeeded() {
+        let copiesInData = data.count / originalCount
+        let maxCopiesAllowed = 5
+
+        guard copiesInData > maxCopiesAllowed else { return }
+
+        // If user is at least 2 copies in from the front, we can drop 1 copy from the front.
+        while index >= 2 * originalCount,
+              data.count / originalCount > maxCopiesAllowed {
+            data.removeFirst(originalCount)
+            index -= originalCount
+        }
+
+        // If user is at least 2 copies from the end, we can drop 1 copy from the end.
+        while index < data.count - 2 * originalCount,
+              data.count / originalCount > maxCopiesAllowed {
+            data.removeLast(originalCount)
+        }
+    }
+
+    // Identify which copy indices we have
+    private func lowestCopyIndex() -> Int {
+        guard let firstID = data.first?.id else { return 0 }
+        return firstID / originalCount
+    }
+
+    private func highestCopyIndex() -> Int {
+        guard let lastID = data.last?.id else { return 0 }
+        return lastID / originalCount
+    }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+struct PageControlView: View {
+
+    let originalCount: Int
+    let pageControl: DisplayablePageControl
+    @Binding var currentIndex: Int
+
+    @State private var localCurrentIndex: Int = 0
+
+    var activeIndicator: DisplayablePageControlIndicator {
+        pageControl.active
+    }
+
+    var indicator: DisplayablePageControlIndicator {
+        pageControl.default
+    }
+
+    var body: some View {
+        if self.originalCount <= 1 {
+            EmptyView()
+        } else {
```

Comment:
Perhaps we don't need the `EmptyView()`
```suggestion
        if self.originalCount > 1 {
```